### DISCUSSION TITLE
Where does A/B swapping logic reside?

### DISCUSSION THREAD

**Discussion by Koseng** [kairos-io/kairos#1638] [Q&A] [answered]
Hi guys,

I try to fully understand the kairos boot process. My understanding so far:

`BIOS -> GRUB -> Image is chainloaded from the GRUB bootloader -> Image is booted directly.`

But how does the A/B swapping work? Where and how is the decision made, which image is chainloaded? How does that work?

Thanks a lot.
Heiko

**Comment by Itxaka** [comment, marked as answer]
Hey @Koseng !

You can check our boot assessment config here: https://github.com/kairos-io/kairos/blob/master/overlay/files/system/oem/08_grub.yaml

It mainly makes it so after an upgrade, if we boot to fail from the active image, we reboot and try to boot from the passive one.

As Kairos is an immutable system, the only way to broke the system is to have a broken upgrade, you cannot break the currently running system unless you go out of your way to do so (enabling debug RW mode for example) so that is the only moment the boot assessment comes into play.

So if we are unable to boot active (A) after an upgrade, it will automatically reboot and set the default boot to the passive image (B).

Notice that we also have the recovery image, which resides in a completely different partition.

**Reply by Koseng** [reply]
Thanks a lot.
